### ISSUE TITLE
Push on nghttpx

### ISSUE THREAD

**Issue by acesso** [tatsuhiro-t/nghttp2#149]
Hello Tatsuhrio-t,

Sorry for opening this as a issue request but I could not find any other mean to contact you (twiter is blocked on my network) and github no longer have any PM feature. 

I´ve forked your project in order to to make a few tests and try to add some sort of server push funcionality on nghttpx by parsing the backend response data before send to frontend. In some point it could be able to push content if the backend reply with a corresponding 2xx and data, this should be true in most cases if nghttpx is running as default mode (reverse proxy).

I know that you added Python API for those kind of situations but due my lack of knowledge in Python I could not use it as a proxy only as a server for static files.

I´m not a C++ programer either but I really apretiate your inputs or guidelines on my idea.

Thanks

**Comment by tatsuhiro-t** [comment]
There is an idea to use X-Associated-Content header field (see #73), which I remember correctly, mod_spdy is using for SPDY push.  Backend server is configured to send back X-Associated-Content header field having URIs or path to resource to push.  nghttpx can parse this header field and initiate the push.  To initiate a push, use nghttp2_submit_push_promise() to tell client we are going to push a content. It just submit PUSH_PROMISE frame and not actual resources.
It would be easier to create Downstream object for each push, then we can use existing infrastructure to deal with their life time.
After PUSH_PROMISE was sent (you can know this by implementing nghttp2_on_frame_send_callback), create DownstreamConnection object and start requesting resource to backend.
After you get at least response headers from the backend, use nghttp2_submit_response() as usual.

**Comment by acesso** [comment]
Thanks for your fast and precise reply ;D

In your proposed implementation, based on X-Associated-Content, it espect this header to be send from backend server, right ? in my case it would require modifications on backend server to send them (i´m using an unmodified lighttpd with http1.1 as backend, no spdy so far). 
The idea was to parse the raw html send from backend server and figure out which resources we can pull from it(using http1.1) and push them to client using http2.

I´m not sure but the nghttpd seems to do something like that by parsing html when using -p option.

**Comment by tatsuhiro-t** [comment]
Web server typically has an option to set arbitrary header field in its configuration.  No source code modification is required in that case.
nghttpd does not parse HTML.  It just push resource if the request path matches the configured path.

**Comment by acesso** [comment]
Quite right, inclusive there more than one way to do this on lighttpd (mod_magnet, mod_setenv and directly from fcgi backend).
creating a new testbed for dev adn trials.

Thanks once again

**Comment by acesso** [comment]
I´m just happy to see the nghttpx updated posted today with push support, using X-Associated-Content will not be suported ? actually link header seems a more standarized way to provide this function.

Testing your server pushes, lots of 304 responses on chrome logs and link headers :+1: 

You should consider updating README.rst to add the now suported feature on nghttpx ;D

**Comment by tatsuhiro-t** [comment]
Yeah, we talked about header field to use and decided to Link header field.  It fits very fine for now.
We updated nghttpx manual page to include how to use server push: https://nghttp2.org/documentation/nghttpx.1.html#server-push

**Comment by acesso** [comment]
Not sure if this is already enabled on master branch or only on nghttp2.org server, but after backend server sends a Link header it does not seems to push.

the link header send was(as seens in -LINFO and on browser debug after nghttpx):

link: </light_button.png>; rel=preload</light_logo.png>; rel=preload</poweredby.png>; rel=preload</favicon.ico>; rel=preload

The provided documentation show only to use one per header and no value for as= (not sure if mandatory by rfc5988, I still need to read it better).

There will be any priorization indication as like X-Associated-Content ?

Thanks

**Comment by tatsuhiro-t** [comment]
To put multiple push resources in one Link header, use "," as delimiter like this:
```
link: </light_button.png>; rel=preload, </light_logo.png>; rel=preload, </poweredby.png>; rel=preload, </favicon.ico>; rel=preload
```

For nginx configuration, following sample should work:

```
add_header link '</assets/css/main.css>; rel=preload';
```

I think "as" is extension parameter and not defined in RFC 5988.  Currently nghttpx just ignores it.
The prioritization in HTTP/2 is not just a numeric value, so the way X-Associated-Content provided does not work anymore.  Pushed resources depend on the associated resource and has default weight.  Client may change the priority using PRIORITY frame.  There is an opinion that server should put higher priority for stylesheets (which can be recognized using as=stylesheet parameter).  But we have not go that far yet.

**Comment by acesso** [comment]
Is there any chance of nghttpx server push be implemented in -s, --http2-bridge mode ?

or

Why does default mode does not use Host header in backend connections (not sure if it should) ?

Motivation:

As I´ve double readed the recent documentation and made the proper tests it seems that default mode (Reverse proxy) does not use Host header on backend requests, and relies only on one or multiple backends defined by -b. In this case nghttpx will only proxy the requests to a single host directly so, if the backend server accepts multiple hosts (virtual hosts) they will not be replied to proxy (due the lack of host header).

By the other hand using -s, --http2-proxy(secure proxy mode) works nicely since it does use Host header when query the backend, but does not handle any link headers provided by backend(yet).

Another thing I´ve noticed is that your README.rst still report nghttpx as NOT capable of push at line 464, seems outdated info ;). 

BTW pushing multiple resources in one link header works fine, and I´ve see the priorization for stylesheets in the code, seems quite sinple to add others if needed.

Thanks

**Comment by tatsuhiro-t** [comment]
> Is there any chance of nghttpx server push be implemented in -s, --http2-bridge mode ?

Planned, but no ETA at the moment.  Probably, we will passthrough PUSH_PROMISE from backend to frontend, rather than checking link header field in this case.

> Why does default mode does not use Host header in backend connections (not sure if it should) ?

Previously, nghttpx does not rewrite Host header field.  But it turns out that it seems standard practice to rewrite Host header field on reverse proxying, at least nginx does that.
If rewriting Host header field is not desirable, use `--no-host-rewrite` option.

I'll fix README.rst, thanks!

**Comment by acesso** [comment]
Sorry, I´ve messed the first question, the option is -s aka http2-proxy(HTTP/1.1 backend) and not http2-bridge(HTTP2 TLS, wich makes sense to passthrough PUSH_PROMISE).

I did not attempted to --no-host-rewrite option, its much more cleary for me now.

**Comment by tatsuhiro-t** [comment]
With `-s` option, we have to limit the number of connections to backend proxy per origin server, since we are going to make HTTP/1 link.  Currently, 8 connections max for each host.
Server push speculatively pushes resources to client and this is not that good for this low connection limit.  On the other hand, reverse proxy case (default mode) can make relatively large number of connections to backend, and no problem for push.

**Comment by gdrouet** [comment]
Hello,

I like the idea of a push initiated by a proxy which intercepts the Link header. However I have a question about rel=preload.

Currently the W3C draft on server-hint (https://w3c.github.io/resource-hints/) defines "preload" as a property that characterize a resource not required to display the current page but another page the user might select. So "preload" resources should be loaded with a low priority. With HTTP/2, it seems it could be fine if the browser decides to load the resource later, taking advantage from multiplexing What do you think?

It seems that currently the draft does not already define a resource that should be pushed because it will be loaded as soon as possible by the loaded page. However, chrome already defined it as "subresource" (https://www.chromium.org/spdy/link-headers-and-server-hint/link-rel-subresource), which seems to be a more appropriated rel the proxy should check to know if the resource needs to be pushed or not.

What do you think?

**Comment by tatsuhiro-t** [comment]
I think preload itself is ok to use current navigation if it is used without "next" load policy.  The W3C draft says that if "next" load policy is used with preload, it means that the resource is used in the next navigation, not the current one.
So using preload without "next" load policy is fine for server push, although we do not check load policy at the moment.  CC: @igrigorik

We first considered to use "subresource", but it turns out that reload deprecates it.
See the relevant discussion at https://github.com/h2o/h2o/pull/133#issuecomment-73270342

**Comment by gdrouet** [comment]
Ok thanks for the information! Working with preload is fine but yes, the proxy should check that "next" load policy is missing to perform a server-push.

**Comment by tatsuhiro-t** [comment]
As of 7451a73defc84376807e59420f89eed9686d1512, nghttpx does not push resources with non-empty loadpolicy.  Earlier resource hint draft (http://www.w3.org/TR/resource-hints/#preload) says about empty loadpolicty:

> NOTE
> An empty loadpolicy indicates that the resource is required for the current navigation context, and that the user agent may apply relevant preprocessing to optimize page performance.

But it seems above statements are gone in current editor's copy.
Anyway, resource hint is still moving target, we keep watching this to adopt new scheme.

**Comment by gdrouet** [comment]
Ok thanks for this contribution.

**Comment by igrigorik** [comment]
@tatsuhiro-t @gdrouet a moving target indeed! :) 

There is a refactor underway to simplify the Preload and Resource Hints spec, which will remove the "loadpolicy" attribute entirely - see https://github.com/w3c/preload/pull/17. Once landed, this update will make preload a "user agent must download" directive that can also be used to initiate server push whenever it is present - i.e. no additional checks required for loadpolicy attribute, or anything like it.

**Comment by tatsuhiro-t** [comment]
That'll make preload even more useful.  Thanks!

**Comment by kazuho** [comment]
Regarding the issue I wonder if there are other aspects to consider.

My understanding is that web-site administrators would use the `rel=preload` attribute in the hope that web browsers would fetch the link if it is not already being cached.  However when nghttp2 (or H2O) sees the header, it pushes the link regardless of whether if the resource is cached within the browser.

In other words, the HTTP2 servers do not work in the way the attribute is expected to act.

IMO we should reconsider using the hint to activate server-push, considering the fact that `rel=preload` is an optimization hint; implementing HTTP2 servers in the way that the use of attribute might cause performance degradation (in case a resource that is already being cached in client is pushed) seems to be like a bad idea.

**Comment by gdrouet** [comment]
You're right, I had the same question, you will find some considerations here: http://stackoverflow.com/questions/29352282/do-browser-cancel-server-push-when-resource-is-in-cache

**Comment by tatsuhiro-t** [comment]
Let's make it clear that server push is speculative as described in https://tools.ietf.org/html/draft-ietf-httpbis-http2-17#section-2:

```
   HTTP/2 adds a new interaction mode, whereby a server can push
   responses to a client (Section 8.2).  Server push allows a server to
   speculatively send data to a client that the server anticipates the
   client will need, trading off some network usage against a potential
   latency gain.  The server does this by synthesizing a request, which
   it sends as a PUSH_PROMISE frame.  The server is then able to send a
   response to the synthetic request on a separate stream.
```

The discussion regarding server push and client side cache is regular topic in httpbis mailing list, and the consequence is always that if client does not want that, do RST_STREAM.
Of course, server can disable server push by inspecting request header fields from client; for example, presence of IMS is good indication that the resources are in client cache.
In general, we don't have facility to know which resources are cached in client, and servers just push.

Back to our link rel=preload solution, I agree that it is not perfect and still has a room to improve (resource hint spec itself is still early draft, isn't it?).
If server admin lists every single resources in link rel=preload without considering proxy pushes everything, then we might waste some traffic before we get RST_STREAMs from client (at most 1RTT?).
To decrease those wasted bytes, probably we need dedicated rel attribute for server push, so we can distinguish regular preload and resources for server push.
Coincidentally, I checked resource hint repository today, and found that preload is completely removed from the spec: https://rawgit.com/w3c/resource-hints/prefetch-prerender/index.html
So, we need to keep an eye on it.

**Comment by kazuho** [comment]
@gdrouet @tatsuhiro-t Thank you for the response.

> The discussion regarding server push and client side cache is regular topic in httpbis mailing list, and the consequence is always that if client does not want that, do RST_STREAM.
> Of course, server can disable server push by inspecting request header fields from client; for example, presence of IMS is good indication that the resources are in client cache.
> In general, we don't have facility to know which resources are cached in client, and servers just push.

I think we can all agree on this.

> Back to our link rel=preload solution, I agree that it is not perfect and still has a room to improve (resource hint spec itself is still early draft, isn't it?).
> If server admin lists every single resources in link rel=preload without considering proxy pushes everything, then we might waste some traffic before we get RST_STREAMs from client (at most 1RTT?).

Yes.  Required time until the load completes will become longer at most 1RTT (note: considering the slow-start, it is likely to become less than 1RTT).  The number of wasted octets would be basically the size of the sendbuf on the server side at the moment it pushes the contents (note: most OSes tend to auto-tune the sendbuf size to fully utilize the bandwidth by observing the RTT).

> To decrease those wasted bytes, probably we need dedicated rel attribute for server push, so we can distinguish regular preload and resources for server push.

Agreed.  IMO that is what we should have done.

> Coincidentally, I checked resource hint repository today, and found that preload is completely removed from the spec: https://rawgit.com/w3c/resource-hints/prefetch-prerender/index.html

Wow!  If `preload` gets removed from the spec, we can continue using the name, since the HTTP/2 servers would be the only ones that recognizes the attribute :-)

**Comment by igrigorik** [comment]
@kazuho @tatsuhiro-t preload has been moved into a separate spec: http://w3c.github.io/preload/

FWIW, I think using preload to initiate push is fine: it has "must fetch" semantics for the user agent; the site is making an explicit request to fetch this resource; the UA can control how push is used: limit number of streams, apply flow control, disable it outright. Further, as Simon mentioned in his [SO reply](http://stackoverflow.com/questions/29352282/do-browser-cancel-server-push-when-resource-is-in-cache), the client can RST_STREAM a resource that is already in cache.

Also, note that the site can specify the preload relation via markup _or_ Link header.. I presume h2o/nghttp2 do not scan the response body - right? As such, if the developer wants the UA to initiate the fetch, they could provide it via markup only; whereas presence of a Link header can be an indication to the server that it should (a) fetch this resource into its own cache, (b) push it to the client.

**Comment by tatsuhiro-t** [comment]
Thanks you for updates @igrigorik! So we are in the right direction with link header fields with preload relation. Yes, we don't parse HTML in proxy.  I think @kazuho wrote about the added complexities if we do and not feasible.  I saw some a project do this though, which is interesting...

**Comment by gdrouet** [comment]
Hi,

That's what I'm personnally working on. My java project is a library that helps the developer to manage his statics and let the container optimize and serve them to the client. So mixing it with a proxy means that the back-end has the knowledge of what should be pushed and has the capacity to give right orders to the proxy with the required headers in the HTTP response.

In my project (http://wuic.github.io/), I have a filter that parses HTML, put the result in a cache and then use collected information to know what should be pushed (https://github.com/wuic/wuic/blob/wuic-0.5.x-snapshot/wuic/servlet/src/main/java/com/github/wuic/servlet/HtmlParserFilter.java#L311-L319).

I love the idea of checking markup in the HTML to distinguish what should be pushed or not. I've started to implement it, but this is a draft that will live with the W3C spec and the considerations of this great discussion.

Thanks.

**Comment by gdrouet** [comment]
@igrigorik in the preload examples I see that in addition to headers, <link> tag could be added to the HTML document. Do you confirm that the <link> tag should be added on top of the document (typically in the <head>) to be fetched ASAP?

Thanks!

**Comment by igrigorik** [comment]
@gdrouet yes, you can use the HTTP header **or** the `<link>` tag to achieve the same results, they are equivalent.

**Comment by gdrouet** [comment]
@igrigorik thanks!

**Comment by tatsuhiro-t** [comment]
Server push from HTTP/2 backend (--http2-bridge) is now supported in master.  Closing this issue.
